ledger/ledger-api-bench-tool/src/main/scala/com/daml/ledger/api/benchtool/submission/CommandSubmitter.scala

```diff
@@ -91,65 +99,90 @@ case class CommandSubmitter(services: LedgerApiServices) {
+      commands = commands,
       workflowId = workflowId,
     )
-    services.commandService.submitAndWait(commands).map(_ => ())
+    services.commandService.submitAndWait(result).map(_ => ())
   }
 
   private def submitCommands(
       descriptor: ContractSetDescriptor,
       signatory: Primitive.Party,
       observers: List[Primitive.Party],
       maxInFlightCommands: Int,
+      submissionBatchSize: Int,
   )(implicit
       ec: ExecutionContext
   ): Future[Unit] = {
     implicit val resourceContext: ResourceContext = ResourceContext(ec)
 
+    val numBatches: Int = descriptor.numberOfInstances / submissionBatchSize
     val progressMeter = CommandSubmitter.ProgressMeter(descriptor.numberOfInstances)
     // Output a log line roughly once per 10% progress, or once every 500 submissions (whichever comes first)
     val progressLogInterval = math.min(descriptor.numberOfInstances / 10 + 1, 500)
-    val progressLoggingSink =
+    val progressLoggingSink = {
+      var lastInterval = 0
       Sink.foreach[Int](index =>
-        if (index % progressLogInterval == 0) {
+        if (index / progressLogInterval != lastInterval) {
+          lastInterval = index / progressLogInterval
           logger.info(progressMeter.getProgress(index))
         }
       )
-    val generator =
-      new CommandGenerator(RandomnessProvider.Default, descriptor, signatory, observers)
 
-    logger.info("Submitting commands...")
+    }
+
+    val generator = new CommandGenerator(
+      randomnessProvider = RandomnessProvider.Default,
+      signatory = signatory,
+      descriptor = descriptor,
+      observers = observers,
+    )
+
+    logger.info(
+      s"Submitting commands ($numBatches commands, $submissionBatchSize contracts per command)..."
+    )
     materializerOwner()
       .use { implicit materializer =>
-        Source
-          .fromIterator(() => (1 to descriptor.numberOfInstances).iterator)
-          .mapAsync(maxInFlightCommands) { index =>
-            generator.next() match {
-              case Success(command) =>
-                submitAndWait(
-                  id = commandId(index),
-                  party = signatory,
-                  command = command(signatory),
-                ).map(_ => index)
-              case Failure(ex) =>
-                Future.failed {
-                  logger.error(s"Command generation failed. Details: ${ex.getLocalizedMessage}", ex)
-                  CommandSubmitter.CommandSubmitterError(ex.getLocalizedMessage)
+        for {
+          _ <- Source
+            .fromIterator(() => (1 to descriptor.numberOfInstances).iterator)
+            .wireTap(i => if (i == 1) progressMeter.start())
+            .mapAsync(32)(index =>
+              Future.fromTry(
+                generator.next().map(cmd => index -> cmd)
+              )
+            )
+            .groupedWithin(submissionBatchSize, 1.minute)
+            .map(cmds => cmds.head._1 -> cmds.map(e => e._2).toList)
```

Comment:
nit
```suggestion
            .map(cmds => cmds.head._1 -> cmds.map(_._2).toList)
```

Reply:
And possibly add a type description to make reasoning easier:
```suggestion
            .map((cmds: Seq[(Int, Command)]) => cmds.head._1 -> cmds.map(_._2).toList)
```